### ISSUE TITLE
high cpu use

### ISSUE THREAD

**Issue by francisdb** [node-schedule/node-schedule#333]
We see long periods of 100% cpu when scheduling jobs or a lot of cron schedules happen at the same time.

We could try to group similar cron jobs on our side but I wonder if this could be optimised on the side of the library.

Some test results below

**setup**
all tests run on my macbook 2,7 GHz Intel Core i7
since node can only use one core 'as fast as possible' means at 100% single core

**results**
creating 10000 empty cron jobs takes ~ 20 seconds 90 mb memory
creating 20000 empty cron jobs takes ~ 130 seconds 130 mb memory
creating 30000 empty cron jobs takes ~ 394 seconds 190 mb memory

running 10000 jobs scheduled at the same time takes 2 minutes even if they only print a line to the console (cpu peaks)

note the non-linear growth in time needed to create jobs (this is also a problem for us as we need to rebuild cron jobs on a restart)

for reference creating 30000 settimeouts takes 0.171 seconds
running 30000 timeouts takes less than a second

```js
const schedule = require('node-schedule')
const util = require('util')
const number = 30000
util.log(`Scheduling ${number} crons`)
const start = new Date().getTime()
for (let i = 0; i < number; i++) {
  const nr = i
  console.log(`create ${nr}`)
  // setTimeout(function(){
  //   console.log(`run ${nr}`)
  // }, 6000)
  schedule.scheduleJob('00 01 23 * * *', function(){
    console.log(`run ${nr}`)
  })
}
const end = new Date().getTime()
util.log(`Ok in ${end-start} ms`)

let p = new Date().getTime()
setInterval(function () {
  p = p + 1000
  const difference = new Date().getTime() - p
  util.log("Difference between expected & actual: " + difference)
}, 1000)
```

**Comment by santigimeno** [comment]
@francisdb I have a hunch that the issue could be in this [tight loop](https://github.com/harrisiirak/cron-parser/blob/1.1.0/lib/expression.js#L365-L462). If that's the case, we could try executing every iteration in a different tick by wrapping the code with `nextTick`.

**Comment by francisdb** [comment]
for our usecase 95% of crons are */15 min for any hour, would this translate into a huge amount o iterations?

**Comment by santigimeno** [comment]
> If that's the case, we could try executing every iteration in a different tick by wrapping the code with nextTick.

Disregard this. I don't think this will work as every operation is synchronous.
TBH, I don't know if there's a proper solution with the current implementation in `cron-parser`.

**Comment by francisdb** [comment]
We had to switch to https://github.com/foreverjs/forever which does not seem to be having this high cpu use (creating 100000 cron jobs in 8s)
For my part this ticket can be closed

**Comment by santigimeno** [comment]
@francisdb thanks for following up. I'll keep the issue opened to see if this can be improved.
Cheers

**Comment by santigimeno** [comment]
After instrumenting the example, the problem is not with `cron-parser` but with: https://github.com/node-schedule/node-schedule/blob/master/lib/schedule.js#L446

**Comment by santigimeno** [comment]
With https://github.com/node-schedule/node-schedule/pull/335, the `30000` jobs take ~ `9700ms` whereas with current master take ~ `322045ms`.

**Comment by francisdb** [comment]
Perfect, thanks for looking into this!

**Comment by santigimeno** [comment]
BTW, fixed in `1.2.1`.
